httpclient5/src/main/java/org/apache/hc/client5/http/entity/LazyDecompressingInputStream.java

```diff
@@ -83,19 +83,29 @@ public boolean markSupported() {
 
     @Override
     public int available() throws IOException {
-        initWrapper();
-        return wrapperStream.available();
+        return initWrapper().available();
     }
 
     @Override
     public void close() throws IOException {
         try {
-            if (wrapperStream != null) {
-                wrapperStream.close();
-            }
+            Closer.close(wrapperStream);
         } finally {
-            wrappedStream.close();
+            super.close();
+        }
+    }
+
+    @Override
+    public synchronized void mark(final int readlimit) {
```

Comment:
Shouldn't we use ReentrantLock instead?

Reply:
@garydgregory No more `synchronized` for us going forward. Thread synchronization must be implemented using locks in order to be compatible with Java 21 fibers

Reply:
Hm, I guess the whole class isn't thread-safe to start with due to `initWrapper()`. I'll rework the PR.

These `synchronized` come from the Java 8, 11, and 17 contracts.

Reply:
Thank you both for the reviews.

The simplest change and implementation that keeps in line with what to have today is to remove the `synchronized` keyword and let the delegate's implementation do what it will. If we wanted mark/reset to be thread-safe, we could make the whole class thread-safe like this:

```java
   private final ReentrantLock lock = new ReentrantLock();
   ...
    /**
     * Thread-safe initialization of the wrapped stream.
     *
     * @return the wrapper stream.
     * @throws IOException if an IO error occurs.
     */
    private InputStream initWrapper() throws IOException {
        if (wrapperStream == null) {
            lock.lock();
            try {
                wrapperStream = wrapperStream != null ? wrapperStream : inputStreamFactory.create(in);
            } finally {
                lock.unlock();
            }
        }
        return wrapperStream;
    }
```
but then each call to the stream would lock-unlock.

So, either:
- Don't use `synchronized` and that's it.
- Only have mark/reset use lock-unlock, or
- Implement using the above thread-safe `initWrapper()`?

For now, the PR takes the simplest route.
WDYT?

Reply:
> * Don't use `synchronized` and that's it.

I would go for this one.